From c37d4646aaa1f3978313c6dd0b0ffa0b505929e5 Mon Sep 17 00:00:00 2001
From: Terry Zhou <bjzhou@marvell.com>
Date: Tue, 17 May 2016 09:56:01 +0800
Subject: [PATCH 394/538] clk: mvebu: corediv: enable using dynamic main PLL
 for Msys

Because main PLL clock feeding coredivclk is defined by SAR settings,
hitherto method of obtaining parent clock from fixed one can't work with
Msys SoC's.

Change-Id: Ic0132f315339c94d75078d1daf3c4fbe8e1ebf2b
Signed-off-by: Terry Zhou <bjzhou@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/29742
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Wilson Ding <dingwei@marvell.com>
Reviewed-by: Lior Amsalem <alior@marvell.com>
---
 drivers/clk/mvebu/clk-corediv.c | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/drivers/clk/mvebu/clk-corediv.c b/drivers/clk/mvebu/clk-corediv.c
index f364760..8125b60 100644
--- a/drivers/clk/mvebu/clk-corediv.c
+++ b/drivers/clk/mvebu/clk-corediv.c
@@ -12,6 +12,7 @@
 
 #include <linux/kernel.h>
 #include <linux/clk-provider.h>
+#include <linux/clk.h>
 #include <linux/of_address.h>
 #include <linux/slab.h>
 #include <linux/delay.h>
@@ -251,6 +252,7 @@ mvebu_corediv_clk_init(struct device_node *node,
 	struct clk_init_data init;
 	struct clk_corediv *corediv;
 	struct clk **clks;
+	struct clk *clk;
 	void __iomem *base;
 	const char *parent_name;
 	const char *clk_name;
@@ -260,7 +262,15 @@ mvebu_corediv_clk_init(struct device_node *node,
 	if (WARN_ON(!base))
 		return;
 
-	parent_name = of_clk_get_parent_name(node, 0);
+	/* Msys SoC's do not use fixed PLL as parent clock */
+	if (of_device_is_compatible(node, "marvell,msys-corediv-clock")) {
+		clk = of_clk_get(node, 0);
+		if (IS_ERR(clk))
+			goto err_unmap;
+		parent_name = __clk_get_name(clk);
+		clk_put(clk);
+	} else
+		parent_name = of_clk_get_parent_name(node, 0);
 
 	clk_data.clk_num = soc_desc->ndescs;
 
-- 
1.9.1

